src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,164 @@
+                elementArray = grow(elementArray, newArraySize);
+            }
+            for (int i = 0; i < list.size(); i++) {
+                elementArray[lastArrayIndex] = list.get(i);
+                lastArrayIndex++;
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (arrayIncludeIndex(index)) {
+            return (T) elementArray[index];
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index "
+                    + index
+                    + " out of bounds for length "
+                    + (lastArrayIndex - 1)
+            );
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (arrayIncludeIndex(index)) {
+            elementArray[index] = value;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index "
+                    + index
+                    + " out of bounds for length "
+                    + (lastArrayIndex - 1)
+            );
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Object temp;
+        if (arrayIncludeIndex(index)) {
+            temp = elementArray[index];
+            elementArray = removeIndex(elementArray, index);
+            lastArrayIndex--;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index "
+                    + index
+                    + " out of bounds for length "
+                    + (lastArrayIndex - 1)
+            );
+        }
+        return (T) temp;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        Object temp;
+        for (int i = 0; i < lastArrayIndex; i++) {
+            if (deepElementCheck(i, element)) {
+                temp = elementArray[i];
+                elementArray = removeIndex(elementArray, i);
+                lastArrayIndex--;
+                return (T) temp;
+            } else if (elementArray[i] == null && element == null) {
+                elementArray = removeIndex(elementArray, i);
+                lastArrayIndex--;
+                return null;
+            }
+        }
+        throw new NoSuchElementException("No such element in the array.");
     }
 
     @Override
     public int size() {
-        return 0;
+        return lastArrayIndex;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return lastArrayIndex == 0;
+    }
+
+    private Object[] addValueByIndex(Object[] oldArray, int index, T value) {
+        Object oldTemp = oldArray[index];
+        Object[] newArray = new Object[oldArray.length];
+        System.arraycopy(oldArray, 0, newArray, 0, index);
+        newArray[index] = value;
+        newArray[index + 1] = oldTemp;
+        System.arraycopy(oldArray, index, newArray, index + 1, oldArray.length - index - 1);
+        return newArray;
+    }
+
+    private Object[] grow(Object[] oldArray, int newArraySize) {
```

Comment:
question <?> QnA
I removed the parameter `Object[] oldArray` because it's really redundant there, but what about `int newArraySize`
if we understand in advance what is the minimum size of the array we need, why don't we immediately increase it to that size and then already do a typical increase by the coefficient. for example, the `addAll(List<T> list)` method receives the list 1_000_000 as input (my Logic)